File: Flustered Pterodactyl/cocos2dx/gaf_runtimes/cocos2d-x/gaf/GAFSpriteWithAlpha.cpp
#include "GAFSpriteWithAlpha.h"
#include "GAFShaderManager.h"

#include "shaders/CCGLProgram.h"
#include "shaders/CCShaderCache.h"
#include "shaders/ccShaders.h"

static const char * kAlphaFragmentShaderFilename = "pcShader_PositionTextureAlpha_frag.fs";

bool GAFSpriteWithAlpha::initWithTexture(CCTexture2D *pTexture, const CCRect& rect, bool rotated)
{
	if (GAFSprite::initWithTexture(pTexture, rect, rotated))
	{
		for (int i = 0; i < 4; ++i)
		{
			_colorTransform[i]     = 1.0f;
			_colorTransform[i + 4] = 0;
		}
		return ensureShaderValid();
	}
	else
	{
		return false;
	}
}

bool GAFSpriteWithAlpha::ensureShaderValid()
{
	CCGLProgram * program = CCShaderCache::sharedShaderCache()->programForKey(kGAFSpriteWithAlphaShaderProgramCacheKey);
	if (!program)
	{
		program = GAFShaderManager::createWithFragmentFilename(ccPositionTextureColor_vert, kAlphaFragmentShaderFilename);
		if (program)
		{
			program->addAttribute(kCCAttributeNamePosition, kCCVertexAttrib_Position);
			program->addAttribute(kCCAttributeNameColor,    kCCVertexAttrib_Color);
			program->addAttribute(kCCAttributeNameTexCoord, kCCVertexAttrib_TexCoords);
			program->link();
			program->updateUniforms();
			CHECK_GL_ERROR_DEBUG();
			CCShaderCache::sharedShaderCache()->addProgram(program, kGAFSpriteWithAlphaShaderProgramCacheKey);
		}
		else
		{
			CCLOGERROR("Cannot load program for GAFSpriteWithAlpha.");
			CC_SAFE_DELETE(program);
			return false;
		}
	}
	setShaderProgram(program);
	program->use();
	_colorTrasformLocation = (GLuint)glGetUniformLocation(program->getProgram(), "colorTransform");
	if (_colorTrasformLocation <= 0)
	{
		CCLOGERROR("Cannot load program for GAFSpriteWithAlpha.");
		return false;
	}
	return true;
}

void GAFSpriteWithAlpha::draw()
{
	ensureShaderValid();
	GAFSprite::draw();
}

void GAFSpriteWithAlpha::setUniformsForFragmentShader()
{	
	glUniform4fv(_colorTrasformLocation, 2, _colorTransform);
}

void GAFSpriteWithAlpha::setColorTransform(const GLfloat * mults, const GLfloat * offsets)
{
	for (int i = 0; i < 4; ++i)
	{
		_colorTransform[i]     = mults[i];
		_colorTransform[i + 4] = offsets[i];
	}
}

void GAFSpriteWithAlpha::setColorTransform(const GLfloat * colorTransform)
{
	for (int i = 0; i < 8; ++i)
	{
		_colorTransform[i] = colorTransform[i];
	}
}

File: BIBs/cocos2dx/gaf_runtimes/cocos2d-iphone/gaf/GAFSpriteWithAlpha.m
#import "GAFSpriteWithAlpha.h"
#import "GAFSprite_Protected.h"
#import "CCGLProgram+Extensions.h"
#import "ccMacros.h"

static NSString * const kAlphaFragmentShaderFilename = @"pcShader_PositionTextureAlpha_frag.fs";
static NSString * const kGAFSpriteWithAlphaShaderProgramCacheKey = @"kGAFSpriteWithAlphaShaderProgramCache";

@interface GAFSpriteWithAlpha ()

- (CCGLProgram *)programForShader;

@end


@implementation GAFSpriteWithAlpha

#pragma mark -
#pragma mark Properties

- (GLfloat *) getColorTransform
{
	return _colorTransform;
}

- (void) setColorTransformMult:(const GLfloat *) mults offsets:(const GLfloat *) offsets
{
	for (int i = 0; i < 4; ++i)
	{
		_colorTransform[i]     = mults[i];
		_colorTransform[i + 4] = offsets[i];
	}
}

- (void) setColorTransform:(const GLfloat *) colorTransform
{
	for (int i = 0; i < 8; ++i)
	{
		_colorTransform[i] = colorTransform[i];
	}
}

#pragma mark -
#pragma mark Initialization & Release

- (id)initWithTexture:(CCTexture2D *)texture rect:(CGRect)rect rotated:(BOOL)rotated
{
    if ((self = [super initWithTexture:texture rect:rect rotated:rotated]))
    {
		for (int i = 0; i < 4; ++i)
		{
			_colorTransform[i]     = 1.0f;
			_colorTransform[i + 4] = 0;
		}        
        self.shaderProgram = [self programForShader];
    }
    return self;
}

- (void)dealloc
{
    [super dealloc];
}

#pragma mark -
#pragma mark Overriden methods

- (void)setUniformsForFragmentShader
{
	glUniform4fv(_colorTrasformLocation, 2, _colorTransform);
}

#pragma mark -
#pragma mark Public methods

#pragma mark -
#pragma mark Private methods

- (CCGLProgram *)programForShader
{
    CCGLProgram *program = [[CCShaderCache sharedShaderCache] programForKey:kGAFSpriteWithAlphaShaderProgramCacheKey];
    if (program == nil)
    {
        program = [[CCGLProgram alloc] initWithVertexShaderByteArray:ccPositionTextureColor_vert
                                              fragmentShaderFilename:kAlphaFragmentShaderFilename];
        
        if (program != nil)
        {
            [program addAttribute:kCCAttributeNamePosition index:kCCVertexAttrib_Position];
            [program addAttribute:kCCAttributeNameColor index:kCCVertexAttrib_Color];
            [program addAttribute:kCCAttributeNameTexCoord index:kCCVertexAttrib_TexCoords];
            
            [program link];
            [program updateUniforms];
            
            CHECK_GL_ERROR_DEBUG();
            
            [[CCShaderCache sharedShaderCache] addProgram:program forKey:kGAFSpriteWithAlphaShaderProgramCacheKey];
            [program release];
        }
        else
        {
            CCLOGWARN(@"Cannot load program for GAFSpriteWithAlpha.");
            [self release];
            return nil;
        }
    }
    
    [program use];
    
    _colorTrasformLocation = (GLuint)glGetUniformLocation(program->_program, "colorTransform");
    
    if (_colorTrasformLocation <= 0)
    {
        CCLOGWARN(@"Cannot get uniforms for kGAFSpriteWithAlphaShaderProgramCacheKey");
    }
    
    return program;
}

@end